### PR TITLE
PIP-45: Implement CoordinationService on top of MetadataStore

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Added `CoordinationService` abstraction implemented on top of `MetadataStore`. 
 
CoordinationService provides: 
 * Locks management
 * Leader election
 * Unique numbers generator

### CODE REVIEW

**Review by rdhabalia — APPROVED** [review]
great work.. 👍

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/coordination/impl/ResourceLockImpl.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        this.state = State.Valid;
+    }
+
+    @Override
+    public synchronized T getValue() {
+        return value;
+    }
+
+    @Override
+    public synchronized CompletableFuture<Void> updateValue(T newValue) {
```
Comment (eolivelli) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/coordination/impl/ResourceLockImpl.java:61]: Probably we do  not need 'synchonized' here, the real critical section is executed in another thread

**Review by eolivelli — APPROVED** [review]
Great work!

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/coordination/impl/ResourceLockImpl.java (hunk 1 above)
Comment (merlimat) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/coordination/impl/ResourceLockImpl.java:61]: We need the `synchronized` for reading the `version` field

Comment (eolivelli) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/coordination/impl/ResourceLockImpl.java:61]: Right. Thanks for the clarification

**Review by jiazhai — APPROVED** [review]

**Review by sijie — APPROVED** [review]

**Pull request merged**